File: sdk/flutter/lib/src/offering.dart
import 'gen/moth/billing/v1/billing.pb.dart' as pb;
import 'gen/moth/billing/v1/billing.pbenum.dart' as pbe;

/// One purchasable tier as the paywall needs it: enough to render a card and
/// match the native store product.
///
/// Price/period are display + analytics metadata; the native store read stays
/// authoritative for the localized price actually charged. The app never gates
/// on [identifier] — it gates on [entitlements] — but the SDK uses it to drive
/// purchases.
class MothOfferingProduct {
  const MothOfferingProduct({
    required this.identifier,
    this.displayName = '',
    this.appleProductId = '',
    this.googleProductId = '',
    this.billingPeriod = '',
    this.priceAmountMicros = 0,
    this.currency = '',
    this.trialPeriod = '',
    this.introPriceAmountMicros = 0,
    this.introPeriod = '',
    this.entitlements = const [],
    this.sortOrder = 0,
    this.highlighted = false,
  });

  factory MothOfferingProduct.fromProto(pb.OfferingProduct proto) =>
      MothOfferingProduct(
        identifier: proto.identifier,
        displayName: proto.displayName,
        appleProductId: proto.appleProductId,
        googleProductId: proto.googleProductId,
        billingPeriod: proto.billingPeriod,
        priceAmountMicros: proto.priceAmountMicros.toInt(),
        currency: proto.currency,
        trialPeriod: proto.trialPeriod,
        introPriceAmountMicros: proto.introPriceAmountMicros.toInt(),
        introPeriod: proto.introPeriod,
        entitlements: List<String>.unmodifiable(proto.entitlements),
        sortOrder: proto.sortOrder,
        highlighted: proto.highlighted,
      );

  /// Stable moth catalog identifier (e.g. `monthly`).
  final String identifier;
  final String displayName;

  /// Store SKUs; either may be empty when the tier ships on one store only.
  final String appleProductId;
  final String googleProductId;

  /// ISO-8601 period descriptors (e.g. `P1M`); empty when unset.
  final String billingPeriod;

  /// Price in micros of [currency] (1,000,000 = one unit). Display metadata;
  /// the native store price is authoritative.
  final int priceAmountMicros;
  final String currency;

  /// Trial/intro descriptors (display + analytics only).
  final String trialPeriod;
  final int introPriceAmountMicros;
  final String introPeriod;

  /// The stable entitlement identifiers this product grants while active.
  final List<String> entitlements;
  final int sortOrder;

  /// Whether this tier is the paywall's highlighted "most popular" tier.
  final bool highlighted;

  /// Whether this product offers a free trial.
  bool get hasTrial => trialPeriod.isNotEmpty;

  /// The store SKU for [store], or the moth [identifier] when the store SKU
  /// is not set.
  String storeProductId(pbe.Store store) {
    final sku = store == pbe.Store.STORE_APPLE
        ? appleProductId
        : googleProductId;
    return sku.isEmpty ? identifier : sku;
  }
}

/// The ordered set of products a paywall presents — the products sharing an
/// `offering` tag, in sort order. Every project has a default offering.
class MothOffering {
  const MothOffering({
    required this.identifier,
    this.isDefault = false,
    this.products = const [],
  });

  factory MothOffering.fromProto(pb.Offering proto) => MothOffering(
    identifier: proto.identifier,
    isDefault: proto.isDefault,
    products: proto.products
        .map(MothOfferingProduct.fromProto)
        .toList(growable: false),
  );

  /// Offering tag; `default` for the project's default offering.
  final String identifier;
  final bool isDefault;

  /// The products to display, in paywall order.
  final List<MothOfferingProduct> products;

  /// True when there is nothing to sell.
  bool get isEmpty => products.isEmpty;

  /// The product with [identifier], or null.
  MothOfferingProduct? productById(String identifier) =>
      products.where((p) => p.identifier == identifier).firstOrNull;

  /// Whether any product in this offering grants [entitlement].
  bool grants(String entitlement) =>
      products.any((p) => p.entitlements.contains(entitlement));
}

/// The rendering variant the paywall screen uses; the token space
/// (colors/spacing/radius) always comes from the theme.
enum MothPaywallLayout {
  /// One card per tier, side by side (the default).
  tiles,

  /// Tiers stacked as full-width rows.
  list,

  /// A single selected tier with a period toggle.
  compact;

  static MothPaywallLayout fromProto(pbe.PaywallLayout proto) =>
      switch (proto) {
        pbe.PaywallLayout.PAYWALL_LAYOUT_LIST => MothPaywallLayout.list,
        pbe.PaywallLayout.PAYWALL_LAYOUT_COMPACT => MothPaywallLayout.compact,
        _ => MothPaywallLayout.tiles,
      };

  static MothPaywallLayout fromName(String name) => switch (name) {
    'list' => MothPaywallLayout.list,
    'compact' => MothPaywallLayout.compact,
    _ => MothPaywallLayout.tiles,
  };
}

/// The public, render-ready paywall configuration, from
/// `moth.billing.v1.GetPaywall`. Copy and layout only — colors/typography
/// inherit from the [MothTheme].
class MothPaywall {
  const MothPaywall({
    this.revisionId = '',
    this.headline = 'Unlock Premium',
    this.subtitle = '',
    this.benefits = const [],
    this.offering = '',
    this.highlightedProductIdentifier = '',
    this.layout = MothPaywallLayout.tiles,
    this.termsUrl,
    this.privacyUrl,
  });

  factory MothPaywall.fromProto(pb.Paywall proto) {
    String? blank(String s) => s.isEmpty ? null : s;
    return MothPaywall(
      revisionId: proto.revisionId,
      headline: proto.headline,
      subtitle: proto.subtitle,
      benefits: List<String>.unmodifiable(proto.benefits),
      offering: proto.offering,
      highlightedProductIdentifier: proto.highlightedProductIdentifier,
      layout: MothPaywallLayout.fromProto(proto.layout),
      termsUrl: blank(proto.termsUrl),
      privacyUrl: blank(proto.privacyUrl),
    );
  }

  /// Rebuilds a paywall persisted with [toJson] (the client-side revision
  /// cache).
  factory MothPaywall.fromJson(Map<String, Object?> json) {
    String? blank(Object? s) => (s is String && s.isNotEmpty) ? s : null;
    return MothPaywall(
      revisionId: json['revisionId'] as String? ?? '',
      headline: json['headline'] as String? ?? '',
      subtitle: json['subtitle'] as String? ?? '',
      benefits: List<String>.unmodifiable(
        (json['benefits'] as List<Object?>? ?? const []).cast<String>(),
      ),
      offering: json['offering'] as String? ?? '',
      highlightedProductIdentifier:
          json['highlightedProductIdentifier'] as String? ?? '',
      layout: MothPaywallLayout.fromName(json['layout'] as String? ?? 'tiles'),
      termsUrl: blank(json['termsUrl']),
      privacyUrl: blank(json['privacyUrl']),
    );
  }

  /// Identifies this version of the paywall config; changes on every admin
  /// edit. Cache the paywall keyed by this value and echo it as
  /// `knownPaywallRevision`.
  final String revisionId;
  final String headline;
  final String subtitle;

  /// Feature/benefit bullets, in display order.
  final List<String> benefits;

  /// The offering tag whose products this paywall lists; empty selects the
  /// default offering.
  final String offering;

  /// The product identifier to render as "most popular"; empty for none.
  final String highlightedProductIdentifier;
  final MothPaywallLayout layout;

  /// Optional legal links rendered in the paywall footer.
  final String? termsUrl;
  final String? privacyUrl;

  /// Serializes the config for the client-side revision cache.
  Map<String, Object?> toJson() => {
    'revisionId': revisionId,
    'headline': headline,
    'subtitle': subtitle,
    'benefits': benefits,
    'offering': offering,
    'highlightedProductIdentifier': highlightedProductIdentifier,
    'layout': layout.name,
    if (termsUrl != null) 'termsUrl': termsUrl,
    if (privacyUrl != null) 'privacyUrl': privacyUrl,
  };
}

File: sdk/react/src/core/offering.ts
import {
  PaywallLayout,
  type Offering,
  type OfferingProduct,
  type Paywall,
} from '../gen/moth/billing/v1/billing_pb.js'

/**
 * One purchasable tier as the paywall needs it: enough to render a card and
 * drive a Stripe Checkout purchase. Price/period are display + analytics
 * metadata. The app never gates on {@link identifier} — it gates on
 * {@link entitlements} — but the SDK uses it to drive purchases.
 */
export interface MothOfferingProduct {
  /** Stable moth catalog identifier (e.g. `monthly`). */
  identifier: string
  displayName: string
  /** Store SKUs; either may be empty when the tier ships on one store only. */
  appleProductId: string
  googleProductId: string
  /**
   * The Stripe Price backing this tier on the web. Empty means the tier is
   * not purchasable on the web — the paywall renders it unavailable rather
   * than hiding it.
   */
  stripePriceId: string
  /** ISO-8601 period descriptor (e.g. `P1M`); empty when unset. */
  billingPeriod: string
  /** Price in micros of {@link currency} (1,000,000 = one unit). */
  priceAmountMicros: number
  currency: string
  /** Trial/intro descriptors (display + analytics only). */
  trialPeriod: string
  introPriceAmountMicros: number
  introPeriod: string
  /** The stable entitlement identifiers this product grants while active. */
  entitlements: readonly string[]
  sortOrder: number
  /** Whether this tier is the paywall's highlighted "most popular" tier. */
  highlighted: boolean
}

/** Whether this product offers a free trial. */
export function productHasTrial(product: MothOfferingProduct): boolean {
  return product.trialPeriod !== ''
}

export function offeringProductFromProto(
  proto: OfferingProduct,
): MothOfferingProduct {
  return {
    identifier: proto.identifier,
    displayName: proto.displayName,
    appleProductId: proto.appleProductId,
    googleProductId: proto.googleProductId,
    stripePriceId: proto.stripePriceId,
    billingPeriod: proto.billingPeriod,
    priceAmountMicros: Number(proto.priceAmountMicros),
    currency: proto.currency,
    trialPeriod: proto.trialPeriod,
    introPriceAmountMicros: Number(proto.introPriceAmountMicros),
    introPeriod: proto.introPeriod,
    entitlements: [...proto.entitlements],
    sortOrder: proto.sortOrder,
    highlighted: proto.highlighted,
  }
}

/**
 * The ordered set of products a paywall presents — the products sharing an
 * `offering` tag, in sort order. Every project has a default offering.
 */
export class MothOffering {
  /** Offering tag; `default` for the project's default offering. */
  readonly identifier: string
  readonly isDefault: boolean
  /** The products to display, in paywall order. */
  readonly products: readonly MothOfferingProduct[]

  constructor(
    identifier: string,
    isDefault = false,
    products: readonly MothOfferingProduct[] = [],
  ) {
    this.identifier = identifier
    this.isDefault = isDefault
    this.products = products
  }

  static fromProto(proto: Offering): MothOffering {
    return new MothOffering(
      proto.identifier,
      proto.isDefault,
      proto.products.map(offeringProductFromProto),
    )
  }

  /** True when there is nothing to sell. */
  get isEmpty(): boolean {
    return this.products.length === 0
  }

  /** The product with `identifier`, or undefined. */
  productById(identifier: string): MothOfferingProduct | undefined {
    return this.products.find((p) => p.identifier === identifier)
  }

  /** Whether any product in this offering grants `entitlement`. */
  grants(entitlement: string): boolean {
    return this.products.some((p) => p.entitlements.includes(entitlement))
  }
}

/**
 * The rendering variant the paywall screen uses; the token space
 * (colors/spacing/radius) always comes from the theme.
 */
export type MothPaywallLayout = 'tiles' | 'list' | 'compact'

/**
 * The public, render-ready paywall configuration, from
 * `moth.billing.v1.GetPaywall`. Copy and layout only — colors/typography
 * inherit from the theme.
 */
export interface MothPaywall {
  /**
   * Identifies this version of the paywall config; changes on every admin
   * edit. Cached keyed by this value and echoed as `knownPaywallRevision`.
   */
  revisionId: string
  headline: string
  subtitle: string
  /** Feature/benefit bullets, in display order. */
  benefits: readonly string[]
  /** The offering tag whose products this paywall lists; empty = default. */
  offering: string
  /** The product identifier to render as "most popular"; empty for none. */
  highlightedProductIdentifier: string
  layout: MothPaywallLayout
  /** Optional legal links rendered in the paywall footer. */
  termsUrl?: string
  privacyUrl?: string
}

export const emptyPaywall: MothPaywall = {
  revisionId: '',
  headline: '',
  subtitle: '',
  benefits: [],
  offering: '',
  highlightedProductIdentifier: '',
  layout: 'tiles',
}

export function paywallFromProto(proto: Paywall): MothPaywall {
  const paywall: MothPaywall = {
    revisionId: proto.revisionId,
    headline: proto.headline,
    subtitle: proto.subtitle,
    benefits: [...proto.benefits],
    offering: proto.offering,
    highlightedProductIdentifier: proto.highlightedProductIdentifier,
    layout: layoutFromProto(proto.layout),
  }
  if (proto.termsUrl !== '') paywall.termsUrl = proto.termsUrl
  if (proto.privacyUrl !== '') paywall.privacyUrl = proto.privacyUrl
  return paywall
}

function layoutFromProto(proto: PaywallLayout): MothPaywallLayout {
  switch (proto) {
    case PaywallLayout.LIST:
      return 'list'
    case PaywallLayout.COMPACT:
      return 'compact'
    default:
      return 'tiles'
  }
}
